### PR TITLE
Add OAuth refresh support

### PR DESCRIPTION
**Pull request by seanh**

### CODE REVIEW

File: src/sidebar/oauth-auth.js [hunk 1]
```diff
@@ -16,23 +16,62 @@ function auth($http, settings) {
   var accessTokenPromise;
   var tokenUrl = resolve('token', settings.apiUrl);
 
+  // Return a new tokenInfo object from the given tokenUrl endpoint response.
+  function tokenInfoFrom(response) {
+    var data = response.data;
+    return {
+      accessToken:  data.access_token,
+      expiresIn:    data.expires_in,
```
Comment (robertknight) [inline src/sidebar/oauth-auth.js]: In any variables that stores durations, it is helpful to make sure the units are clear (seconds, mills, nano?). This could be either via naming (eg. `expiresInSecs`) or via documentation of the type somewhere (eg. a JSDoc `@typedef`)

Comment (seanh) [inline src/sidebar/oauth-auth.js]: Hmm. `expires_in` is part of the OAuth spec. I did change it to `javaScript` case styling but I think maybe we should stick to the spec's names for things as much as possible? So I think adding a typedef sounds good.

File: src/sidebar/oauth-auth.js [hunk 2]
```diff
@@ -16,23 +16,62 @@ function auth($http, settings) {
...
+      }
+      return tokenInfoFrom(response);
+    });
+  }
+
+  // Exchange the refresh token for a new access token and refresh token pair.
+  // See https://tools.ietf.org/html/rfc6749#section-6
+  function refreshAccessToken(refreshToken) {
+    var data = {grant_type: 'refresh_token', refresh_token: refreshToken};
+    postToTokenUrl(data).then(function (response) {
```
Comment (robertknight) [inline src/sidebar/oauth-auth.js]: If this fails, the error will be silently dropped*. In general, a promise chain should either be returned from the function or rejection should be explicitly handled.

* Not always true - modern browsers have heuristics that they will use to warn about unhandled Promise rejections, but I think I still favour handling them explicitly.

File: src/sidebar/test/oauth-auth-test.js [hunk 3]
```diff
@@ -109,5 +112,100 @@ describe('oauth auth', function () {
...
+            fakeHttp.post,
+            'https://hypothes.is/api/token',
+            expectedBody,
+            {headers: {'Content-Type': 'application/x-www-form-urlencoded'},
+          });
+        };
+      }
+
+      function assertThatTokenGetterNowReturnsNewAccessToken () {
+        return auth.tokenGetter().then(function (token) {
```
Comment (robertknight) [inline src/sidebar/test/oauth-auth-test.js]: `auth.tokenGetter()` will be invoked _when the promise chain is constructed_ not when the previous promise in the chain is resolved, which is I think what you intended to do.

Unless you are going to re-use a `then` handler, I would suggest just inlining this as an anonymous function in the `then` block below, perhaps with a one line comment at the top.

File: src/sidebar/oauth-auth.js [hunk 4]
```diff
@@ -16,23 +16,62 @@ function auth($http, settings) {
...
+    var data = response.data;
+    return {
+      accessToken:  data.access_token,
+      expiresIn:    data.expires_in,
+      refreshToken: data.refresh_token,
+    };
+  }
+
+  // Post the given data to the tokenUrl endpoint as a form submission.
+  // Return a Promise for the POST request.
```
Comment (robertknight) [inline src/sidebar/oauth-auth.js]: This returns a Promise for _the response_, not the request ;) In other words, the convention is to say that something is or returns a Promise _for the thing it will resolve to_.

File: src/sidebar/oauth-auth.js [hunk 5]
```diff
@@ -16,23 +16,62 @@ function auth($http, settings) {
...
+  // Set a timeout to refresh the access token a few minutes before it expires.
+  function refreshAccessTokenBeforeItExpires(tokenInfo) {
+    var delay = tokenInfo.expiresIn * 1000;
+
+    // We actually have to refresh the access token _before_ it expires.
+    // If the access token expires in one hour, this should refresh it in
+    // about 55 mins.
+    delay = Math.floor(delay * 0.91);
+
+    window.setTimeout(refreshAccessToken, delay, tokenInfo.refreshToken);
```
Comment (robertknight) [inline src/sidebar/oauth-auth.js]: Ah. I did not know that `setTimeout` accepted function params for the callback as argument.

File: src/sidebar/test/oauth-auth-test.js [hunk 6]
```diff
@@ -109,5 +112,100 @@ describe('oauth auth', function () {
...
+        .then(resetHttpSpy)
+        .then(assertThatTokenGetterNowReturnsNewAccessToken)
+        .then(expireAccessToken)
+        .then(assertRefreshTokenWasUsed('second_refresh_token'));
+    });
+
+    // While a refresh token HTTP request is in-flight, calls to tokenGetter()
+    // should just return the old access token immediately.
+    it('returns the access token while a refresh is in-flight', function() {
+      return auth.tokenGetter().then(function(first_access_token) {
```
Comment (robertknight) [inline src/sidebar/test/oauth-auth-test.js]: Var names should be camelCased (`first_access_token`)

File: src/sidebar/test/oauth-auth-test.js [hunk 7]
```diff
@@ -109,5 +112,100 @@ describe('oauth auth', function () {
...
+      }
+
+      return auth.tokenGetter()
+        .then(resetHttpSpy) // Reset fakeHttp.post.callCount to 0 so that the
+                            // initial access token request isn't counted.
+        .then(auth.tokenGetter)
+        .then(makeServerUnresponsive)
+        .then(auth.tokenGetter)
+        .then(expireAccessToken)
+        .then(assertThatOnlyOneRefreshRequestWasSent);
```
Comment (robertknight) [inline src/sidebar/test/oauth-auth-test.js]: If you are not re-using the logic for a `then` callback or it is very simple, I'm not convinced that it is clearer to put it in a separate function - as I have to scroll up to find out what the actual assertion is.

File: src/sidebar/oauth-auth.js (hunk 2 above)
Comment (seanh) [inline src/sidebar/oauth-auth.js]: Yeah, I know it will be, but oauth error handling in the client is generally pretty nonexistent, so I want to do a separate PR (probably multiple PRs) for it.
